rfcs/0045-deprecate-url-syntax.md

```diff
@@ -0,0 +1,70 @@
+---
+feature: deprecate_url_syntax
+start-date: 2019-04-28
+author: Michael Raskin
+co-authors: 
+related-issues: 
+---
+
+# Summary
+[summary]: #summary
+
+Gradually deprecate the use of unquoted URL syntax in Nix code and especially
+Nixpkgs.
+
+# Motivation
+[motivation]: #motivation
+
+Nix language has a special syntax for URLs. However, unlike paths, URLs do not
+have any special support that would make the difference useful. Moreover, using
+variable expansion in URLs requires some URLs to be quoted strings anyway. So
+the most consistent approach is to always use quoted strings to represent URLs.
+Additionally, a semicolon immediately after the URL can be mistaken for a part
+of URL by language-agnostic tools such as terminal emulators.
+
+In case of future breaking changes in the Nix syntax, it would be nice to make
+the `x:x` snippet parse to the identity function, and not to an URL. Tools
+targeting only Nixpkgs codebase can ignore URL syntax once Nixpkgs phases out
+its use.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a note in the Nix manual that the special unquoted URL syntax is
+deprecated.
+
+Add a note in the Nixpkgs manual that the unquoted URL syntax is deprecated,
+changes to Nixpkgs should not increase its use, and it is recommended to
+convert URLs to quoted strings when changing them.
```

Comment:
Is it possible to be more specific? How would everybody be notified about this new rule?

Reply:
1. I think the mere fact of an RFC passing is a noticeable effect event for a large fraction of active contributors. (It is visible in this repository, announced via Discourse, gets into the «weekly», is likely to be discussed a bit on IRC etc.)

For this specific RFC just the announcement «RFC on Deprecation of URL syntax passed» conveys enough information.

2. It will be noted in two out of three manuals.

3. Once we have an RFC that something «should not» happen, a request for cosmetic cleanup in a PR (and such requests happen) is more likely to contain a mention of quoting URLs (if relevant for the package in question). This also spreads the knowledge.

I do have an impression that quite a few people mentioned that they are disappointed they cannot refer to a policy that quoted URLs are better, so I expect the review channel of information dissemination to perform well.

Appendix A: future work — maybe the tool gets implemented, then we can open a countdown issue and maybe make ofborg check that PRs do not make things worse).

My plan is indeed just these points. I think that these things do not really need any additions to the text of RFC, and I think these mechanisms will be enough to distribute the information.

If you think that any of these four channels benefit from an addition to RFC text, or that there are other information distribution channels that should be used (and mentioned in the RFC text) please give some details.

Reply:
It doesn't have to be formalized in the RFC but it's good to have an accompanying discussion.

For example you mentioned "they are disappointed they cannot refer to a policy". Would the RFC act as such policy? If not, maybe we should have a list of official policies in place.

Reply:
I think having a deprecation notice in the manual (added in accordance with an explicit bullet point in an approved/accepted RFC) is close enough to official policy for the purpose of asking people not to do the deprecated things.

I didn't start the discussion myself, because I thought that (1) is assumed, (2) and (A) are explicitly mentioned, and (3) is a manual variant of (A) in a sense.